src/lib/fastDownward/analysis.ts

```diff
@@ -0,0 +1,207 @@
+import { LearningUnit, Path, Skill, SkillAnalyzedPath } from "../types";
+import { SearchNode } from "./searchNode";
+import { State } from "./state";
+import { GlobalKnowledge } from "./global-knowledge";
+
+/**
+ * Compute which LearningUnits are reversed reachable based on the given goals in state.
+ * Using teachingGoals of the LearningUnits and the skill groups
+ */
+function availableActions<LU extends LearningUnit>(
+	currentState: State,
+	learningUnits: ReadonlyArray<LU>,
+	globalKnowledge: GlobalKnowledge
+) {
+    // Find the LearningUnits for each goal in the state using teachingGoals.
+    let usefulLus = learningUnits.filter(lu => lu.teachingGoals.some(skill => currentState.learnedSkills.includes(skill.id)));
+    
+    // Find the LearningUnits for each goal in the state using globalKnowledge (Skill groups).
+    globalKnowledge.getAllParents().forEach(parent => {
+        if (currentState.getHashCode().includes(parent.id) ) {
+            const lu = learningUnits.filter(lu => lu.teachingGoals.some(skill => parent.nestedSkills.includes(skill.id)));
+            usefulLus = usefulLus.concat(lu);
+        }
+    });
+    
+    // Remove duplicate LearningUnits
+    return [...new Set(usefulLus)];
+}
+
+/**
+ * Analysis a goal to find the missing skills.
+ * @param goal The skills that should be learned (should not contain skills that are already known).
+ * @param globalKnowledge The set of all skills (independent of what was already learned and what should be learned).
+ * @param learningUnits All learning units of the system to learn new skills.
+ * @returns A list of the missing skills with the sub paths for them. 
+*/
+export function skillAnalysis<LU extends LearningUnit>(
+    globalKnowledge: GlobalKnowledge,
+    learningUnits: ReadonlyArray<LU>,
+    goal: Skill[]
+): SkillAnalyzedPath[] | null {
+
+    const openListMap = new Map();
+    let currentNode: SearchNode<LU>;
+
+    // Create initial state with goals and globalKnowledge
+    const initialState = new State(
+        goal.map(skill => skill.id),
+        globalKnowledge
+    );
+
+    // Remove duplicate LearningUnits
+    initialState.learnedSkills = [...new Set(initialState.learnedSkills)];
+
+    const openList: SearchNode<LU>[] = [new SearchNode<LU>(initialState, null, null, 0, 0)];
+    
+    // Use goalString to find missing skills that does not have a requirement
+    let goalString = `,`.concat(initialState.learnedSkills.join(",").concat(`,`));
+    
+        while (openList.length > 0) {
+        currentNode = openList.shift()!;

```

Comment:
Missing indentation. Prettier Addon should atomatically increase the indentation level here.

Reply:
Solved